rad/prelude/cmd-parsing.rad

```diff
@@ -27,32 +30,60 @@
   (fn [cmd]
     (string-append "Too many arguments for command \"" cmd "\"")))
 
-(def /cmd-1-opts
-  "Matches a command `cmd` with one argument and an optional
-  combination of options that are allowed by opts-map"
-  (fn [cmd opts opts-map]
+(def /cmd-opts
+  "Matches the command `cmd` and a combination of optional parameters, where
+  `opt-match` matches the options to the keys of `multi-ops`. `multi-ops` is a
+  dict of option-keys with multiple arguments. If no options are passed, the
+  `default` is returned."
+  (fn [cmd opts cmd-opts help]
+    (def default
+      (dict-from-seq
+        (map (fn [opt] (list (lookup :key opt) (lookup :default opt))) cmd-opts)))
+    (def-rec parse-cli
+      (fn [args]
+        (def parse-arg
+          (fn [opt args]
+            (def opt-key (lookup :key opt))
+            (match args
+              (/cons 'val 'rest)
+                  (if (elem? val (lookup :possible-values opt))
+                    (modify-map opt-key (fn [x] (cons val x)) (parse-cli rest))
+                    (parse-failure (string-append "Invalid argument " val " for option " (show opt-key) ".") help))
```

Comment:
At the moment `--state --fancy` warns you that `--fancy` is an invalid value for `--state`. It should warn you that `--state` is missing an argument. I think this can be remedied by checking whether `val` starts with `-`.